extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/CodeAuthenticationMechanism.java

```diff
@@ -298,5 +344,66 @@ private void removeCookie(RoutingContext context, TenantConfigContext configCont
                 cookie.setPath(auth.cookiePath.get());
             }
         }
+        return cookie;
+    }
+
+    private boolean isLogout(RoutingContext context, TenantConfigContext configContext) {
+        Optional<String> logoutPath = configContext.oidcConfig.logout.path;
+
+        if (logoutPath.isPresent()) {
+            return context.request().absoluteURI().equals(
+                    buildUri(context, logoutPath.get()));
+        }
+
+        return false;
+    }
+
+    private SecurityIdentity trySilentRefresh(TenantConfigContext configContext, String idToken, String refreshToken,
+            RoutingContext context) {
+        OidcTenantConfig config = configContext.oidcConfig;
+
+        if (config.token.refreshTimeout.isPresent()) {
+            OAuth2AuthProviderImpl auth = ((OAuth2AuthProviderImpl) configContext.auth);
+            JWT jwt = auth.getJWT();
+            JsonObject tokenJson;
+
+            try {
+                tokenJson = jwt.decode(idToken);
+            } catch (Exception cause) {
+                throw new AuthenticationFailedException(cause);
+            }
+
+            Long iat = tokenJson.getLong("iat");
+            long now = System.currentTimeMillis() / 1000;
+            Integer timeout = config.token.refreshTimeout.get();
+
+            if (now - iat >= timeout) {
+                CompletableFuture<SecurityIdentity> cf = new CompletableFuture<>();
+                OAuth2TokenImpl token = new OAuth2TokenImpl(configContext.auth, new JsonObject());
+
+                token.principal().put("refresh_token", refreshToken);
+
+                token.refresh(new Handler<AsyncResult<Void>>() {
+                    @Override
+                    public void handle(AsyncResult<Void> result) {
+                        if (result.succeeded()) {
+                            String rawIdToken = token.principal().getString("id_token");
+                            IdTokenCredential idToken = new IdTokenCredential(rawIdToken, context);
+                            QuarkusSecurityIdentity identity = validateAndCreateIdentity(idToken, config,
+                                    jwt.decode(idToken.getToken()));
+                            // after a successful refresh, rebuild the identity and update the cookie 
+                            processSuccessfulAuthentication(context, configContext, token, identity);
+                            cf.complete(identity);
+                        } else {
+                            cf.completeExceptionally(new AuthenticationFailedException(result.cause()));
```

Comment:
`AuthenticationCompletionException`

Reply:
What is the difference when using one or another? I'm expecting to send a challenge, so `AuthorizationFailedException` is what I need?

Reply:
If the user has authentiicated but the code grant fails then there will be a redirect loop, so we should avoid this here

Reply:
Not the case here, then ? If the refresh fails we invalidate the cookie and re-start the login flow.